File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/ResolvedModel/ParameterValueSet.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.ResolvedModel
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using System.Collections.Generic;

    internal class ParameterValueSet
    {
        internal ParameterCollection PC { get; set; }
        internal List<dynamic> Values { get; set; }
        internal CdmCorpusContext Ctx { get; set; }
        internal int Length
        {
            get
            {
                if (this.PC?.Sequence != null)
                {
                    return this.PC.Sequence.Count;
                }
                return 0;
            }
        }

        internal List<bool> WasSet { get; set; }

        internal ParameterValueSet(CdmCorpusContext ctx, ParameterCollection pc, List<dynamic> values, List<bool> wasSet)
        {
            this.PC = pc;
            this.Values = values;
            this.WasSet = wasSet;
            this.Ctx = ctx;
        }
        public int IndexOf(CdmParameterDefinition paramDef)
        {
            return this.PC.Ordinals[paramDef];
        }
        public CdmParameterDefinition FetchParameterAtIndex(int paramIndex)
        {
            return this.PC.Sequence[paramIndex];
        }
        public dynamic FetchValue(int paramIndex)
        {
            return this.Values[paramIndex];
        }
        public string FetchValueString(ResolveOptions resOpt, int paramIndex)
        {
            return new ParameterValue(this.Ctx, this.PC.Sequence[paramIndex], this.Values[paramIndex]).FetchValueString(resOpt);
        }
        internal ParameterValue FetchParameterValueByName(string paramName)
        {
            int paramIndex = this.PC.FetchParameterIndex(paramName);
            return new ParameterValue(this.Ctx, this.PC.Sequence[paramIndex], this.Values[paramIndex]);
        }

        public void SetParameterValue(ResolveOptions resOpt, string paramName, dynamic value)
        {
            int paramIndex = this.PC.FetchParameterIndex(paramName);
            this.Values[paramIndex] = ParameterValue.FetchReplacementValue(resOpt, this.Values[paramIndex], value, true);
            this.WasSet[paramIndex] = true;
        }

        public ParameterValueSet Copy()
        {
            List<dynamic> copyValues = new List<dynamic>(this.Values);
            List<bool> copyWasSet = new List<bool>(this.WasSet);
            ParameterValueSet copy = new ParameterValueSet(this.Ctx, this.PC, copyValues, copyWasSet);
            return copy;
        }

        public void Spew(ResolveOptions resOpt, StringSpewCatcher to, string indent)
        {
            for (int i = 0; i < this.Length; i++)
            {
                ParameterValue parameterValue = new ParameterValue(this.Ctx, this.PC.Sequence[i], this.Values[i]);
                parameterValue.Spew(resOpt, to, indent + '-');
            }
        }
    }
}

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/resolvedmodel/ParameterValueSet.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.resolvedmodel;

import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusContext;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmParameterDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.StringSpewCatcher;
import com.microsoft.commondatamodel.objectmodel.utilities.ResolveOptions;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * @deprecated This class is extremely likely to be removed in the public interface, and not meant
 * to be called externally at all. Please refrain from using it.
 */
@Deprecated
public class ParameterValueSet {

  private ParameterCollection pc;
  // TODO-BQL Investigate values, and wasSet, they seems to be always use together. Might want to declare as List<Tuple<Object,Boolean>> ?
  private List<Object> values;// TODO-BQ: This may need to be revisit, uncertain if type is correct.
  private CdmCorpusContext ctx;
  private List<Boolean> wasSet;

  ParameterValueSet(
      final CdmCorpusContext ctx,
      final ParameterCollection pc,
      final List<Object> values,
      final List<Boolean> wasSet) {
    this.pc = pc;
    this.values = values;
    this.wasSet = wasSet;
    this.ctx = ctx;
  }

  public int indexOf(final CdmParameterDefinition paramDef) {
    return pc.getOrdinals().get(paramDef);
  }

  public CdmParameterDefinition fetchParameter(final int paramIndex) {
    return pc.getSequence().get(paramIndex);
  }

  public Object fetchValue(final int paramIndex) {
    return values.get(paramIndex);
  }

  public String fetchValueString(final ResolveOptions resOpt, final int paramIndex)
      throws IOException {
    return new ParameterValue(ctx, pc.getSequence().get(paramIndex), values.get(paramIndex))
        .fetchValueString(resOpt);
  }

  /**
   *
   * @param paramName
   * @return
   * @deprecated This function is extremely likely to be removed in the public interface, and not
   * meant to be called externally at all. Please refrain from using it.
   */
  @Deprecated
  public ParameterValue fetchParameterValue(final String paramName) {
    final int paramIndex = pc.fetchParameterIndex(paramName);
    return new ParameterValue(ctx, pc.getSequence().get(paramIndex), values.get(paramIndex));
  }

  public void setParameterValue(final ResolveOptions resOpt, final String paramName, final Object value) {
    final int paramIndex = pc.fetchParameterIndex(paramName);
    values.set(paramIndex,
        ParameterValue.fetchReplacementValue(resOpt, values.get(paramIndex), value, true));
    wasSet.set(paramIndex, true);
  }

  public ParameterValueSet copy() {
    final List<Object> copyValues = new ArrayList<>(values);
    final List<Boolean> copyWasSet = new ArrayList<>(wasSet);
    final ParameterValueSet copy = new ParameterValueSet(ctx, pc, copyValues, copyWasSet);
    return copy;
  }

  public void spew(final ResolveOptions resOpt, final StringSpewCatcher to, final String indent)
      throws IOException {
    for (int i = 0; i < length(); i++) {
      final ParameterValue parameterValue = new ParameterValue(ctx, pc.getSequence().get(i),
          values.get(i));
      parameterValue.spew(resOpt, to, indent + '-');
    }
  }

  ParameterCollection getPc() {
    return pc;
  }

  void setPc(final ParameterCollection pc) {
    this.pc = pc;
  }

  public List<Object> getValues() {
    return values;
  }

  void setValues(final List<Object> values) {
    this.values = values;
  }

  CdmCorpusContext getCtx() {
    return ctx;
  }

  void setCtx(final CdmCorpusContext ctx) {
    this.ctx = ctx;
  }

  List<Boolean> getWasSet() {
    return wasSet;
  }

  void setWasSet(final List<Boolean> wasSet) {
    this.wasSet = wasSet;
  }

  public int length() {
    if (pc != null && pc.getSequence() != null) {
      return pc.getSequence().size();
    }

    return 0;
  }
}
